### ISSUE TITLE
[SUPPORT] - Deprecation Warning for TripleDES Algorithm in Paramiko

### ISSUE THREAD

**Issue by ErickAlejo** [paramiko/paramiko#2419]
### Are you using paramiko as a client or server?

Both

### What feature(s) aren't working right?

SSH

### What version(s) of paramiko are you using?

3.4.0

### What version(s) of Python are you using?

3.10.12

### What operating system and version are you using?

Linux Mint

### What are you trying to do with paramiko?

Description:

I am encountering a deprecation warning when using Paramiko in my Python script. The warning message is as follows:

/home/user/Documents/projects/myproject/lib/python3.10/site-packages/paramiko/transport.py:259: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from this module in 48.0.0.
  "class": algorithms.TripleDES,

Additional Information:
This warning indicates that the TripleDES algorithm has been moved to the cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES module and will be removed in a future release (version 48.0.0) of the cryptography library. It is important to address this deprecation to ensure compatibility with future versions of the cryptography library.

### How are you trying to do it, and what's happening instead?

I try to run my script, but I am only importing netmiko, without connecting via ssh to any computer, just importing

**Comment by alepinio** [comment]
Confirm. The message is quite annoying

```
❯ python3
Python 3.8.10 (default, Nov 22 2023, 10:22:35) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import paramiko
/home/apinio/.local/lib/python3.8/site-packages/paramiko/pkey.py:100: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from this module in 48.0.0.
  "cipher": algorithms.TripleDES,
/home/apinio/.local/lib/python3.8/site-packages/paramiko/transport.py:259: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from this module in 48.0.0.
  "class": algorithms.TripleDES,
>>> print(paramiko.__version__)
3.4.0
```

**Comment by 1001nights** [comment]
Same annoying warnings here.
paramiko 3.4.0
python 3.11

**Comment by bryanbanda** [comment]
Issue seems to be from the underlying `cryptography` library being on `43.0.0` which indeed moved the algorithms to the decrepit algorithms. Works fine if you pin `cryptography < 43.0.0` but I wouldn't want to do pin that.

**Comment by Wedge009** [comment]
Encountered this just now as well. I'm guessing this is because paramiko accommodates TripleDES cipher in public keys:

https://github.com/paramiko/paramiko/blob/51eb55debf2ebfe56f38378005439a029a48225f/paramiko/pkey.py#L85-L104

https://github.com/paramiko/paramiko/blob/51eb55debf2ebfe56f38378005439a029a48225f/paramiko/transport.py#L221-L263

**Comment by OmarHory** [comment]
Any solution?

**Comment by Wedge009** [comment]
I think it depends on how the developers want to work around it. I imagine the immediate resolution is to change the references of `algorithms.TripleDES` to the new location, but the warning also states it will be removed altogether in `cryptography` 48.0.0. So `paramiko` may need to implement its own deprecation handling.

**Comment by intgr** [comment]
> So paramiko may need to implement its own deprecation handling.

It should be considered if it's possible to remove TripleDES support entirely. 3DES has been considered to be cryptographically weak for a long time.

It's 2024: AES is so prevalent, I doubt you can find any SSH server that doesn't support AES, but does support 3DES.

It seems OpenSSH disabled 3DES by default in their `ssh` client in v7.4, released in 2016 https://www.openssh.com/txt/release-7.4

>  * ssh(1): Remove 3des-cbc from the client's default proposal. 64-bit
>   block ciphers are not safe in 2016 and we don't want to wait until
>   attacks like SWEET32 are extended to SSH. As 3des-cbc was the
>   only mandatory cipher in the SSH RFCs, this may cause problems
>   connecting to older devices using the default configuration,
>   but it's highly likely that such devices already need explicit
>   configuration for key exchange and hostkey algorithms already
>   anyway.

**Comment by anashakt** [comment]
I am using python 3.12 and paramiko v3.4.0. Same issue

**Comment by Wedge009** [comment]
> It should be considered if it's possible to remove TripleDES support entirely. 3DES has been considered to be cryptographically weak for a long time.

I agree, but I can't presume to know how the `paramiko` devs want to proceed.

**Comment by antimattergroup** [comment]
I tried every single version from 3.0, they still has this issue.

**Comment by bryanbanda** [comment]
Short-term fix that worked for me:

Add and/or pin `cryptography<43.0.0` in your requirements until paramiko devs decide on a way forward.

**Comment by mushishi7** [comment]
In my environment, the issue was resolved by downgrading cryptography to version 42.0.0. ?

**Comment by zsfc** [comment]
Same issue here (Python 3.12.3, paramiko 3.4.0, cryptography 43.0.0).

If you looking for a quick fix to temporarily suppress the import warning, you could change `import paramiko` to:
```python
import warnings
with warnings.catch_warnings(action="ignore"):
    import paramiko
```

**Comment by ErickAlejo** [comment]
I guess it's time to close this issue, the truth is that everyone's comments have helped me a lot with some projects that had the most recent version of Paramiko

The solutions that have been proposed for anyone who has this problem are the following:

- Use the cryptography library in versions lower than 43.0.0 `"cryptography<43.0.0"`
- 
- Ignore warnings with the 'warning' library  :
```
"import warnings
from cryptography.utils import CryptographyDeprecationWarning
with warnings.catch_warnings(action="ignore", category=CryptographyDeprecationWarning):
import paramiko"
```

Greetings to all, good day and hopefully in the next "commit" the devs will fix this.

**Comment by 1001nights** [comment]
Is that how things are done here? Close an issue just because there are workarounds? Surely it should remain open until the underlying cause has been addressed, and not just in the hope that in the next commit the devs will fix it?
How will the issue be tracked and notified in release notes so that people can remove their workarounds otherwise?

**Comment by matt-lee** [comment]
I agree, why close this issue? 3DES is considered insecure by NIST and OpenSSL, why should paramiko continue supporting it?

**Comment by bskinn** [comment]
Agreed, this should stay open until actively addressed.

**Comment by AdrianFretwell** [comment]
I am also now encountering this issue.  Happy to implement the work around, but would like to see the issue addressed.  I do not use TripleDes anywhere.

**Comment by miteshvermaae** [comment]
this is how i fixed it, by adding the reference to the correct class location
and did it for all the errors.

goto pkey.py file and update the following

1)  **add this in imports** -> from cryptography.hazmat.decrepit.ciphers.algorithms import TripleDES

2) **replace following cipher**
from this ->  
"DES-EDE3-CBC": {
            "cipher": algorithms.TripleDES,
            "keysize": 24,
            "blocksize": 8,
            "mode": modes.CBC,
        },


to this ->
 "DES-EDE3-CBC": {
            "cipher": TripleDES,
            "keysize": 24,
            "blocksize": 8,
            "mode": modes.CBC,
        },

**Comment by codesorcery** [comment]
@miteshvermaae with that approach `paramiko` would only work with `cryptography>=43`, since the import would fail for older version (where the class hasn't been relocated yet). You'd need to do a conditional import based on the version of `cryptopgraphy`.

**Comment by Wedge009** [comment]
#2421

**Comment by fujch7** [comment]
roll back. 
``` shell
pip3 install paramiko==3.0 cryptography==36.0
```

**Comment by isosphere** [comment]
> roll back.
> 
> ```shell
> pip3 install paramiko==3.0 cryptography==36.0
> ```

Running old versions of cryptography is risky; new CVEs get published for it pretty frequently.

**Comment by sebastian-correa** [comment]
Can we ping a maintainer to see if they are aware of the issue/have a statement about it? Or is that considered terrible etiquette for these kinds of issues? I don't want to annoy them, but workarounds aren't super _pretty_.

**Comment by ErickAlejo** [comment]
I put the "support" tag on it because I thought it would be temporary, but I've seen that many people have been affected. Two weeks ago I published this "issue" but nothing official, I guess it's one of the bad things about open-sources :(

Por cierto ¿te afecto en algun proyecto?, en donde trabajo aún no ha  sucedido nada pero porque tome una solución de las qu edieron los demas :v

**Comment by ErickAlejo** [comment]
> Can we ping a maintainer to see if they are aware of the issue/have a statement about it? Or is that considered terrible etiquette for these kinds of issues? I don't want to annoy them, but workarounds aren't super _pretty_.

I put the "support" tag on it because I thought it would be temporary, but I've seen that many people have been affected. Two weeks ago I published this "issue" but nothing official, I guess it's one of the bad things about open-sources :(

A ti ¿te afecto en algun proyecto?, yo por suerte donde trabajo tome una de las medidas que dieron las personas de este issue

**Comment by bskinn** [comment]
There's a PR opened to address this: https://github.com/paramiko/paramiko/pull/2421

Please follow there for updates.

**Comment by sebastian-correa** [comment]
> > Can we ping a maintainer to see if they are aware of the issue/have a statement about it? Or is that considered terrible etiquette for these kinds of issues? I don't want to annoy them, but workarounds aren't super _pretty_.
> 
> I put the "support" tag on it because I thought it would be temporary, but I've seen that many people have been affected. Two weeks ago I published this "issue" but nothing official, I guess it's one of the bad things about open-sources :(
> 
> A ti ¿te afecto en algun proyecto?, yo por suerte donde trabajo tome una de las medidas que dieron las personas de este issue

Didn't break anything in our end for now, but we started getting those warnings.

**Comment by JohnVillalovos** [comment]
As a note. The TripleDES cipher is not being removed from the cryptography library. It is just being removed from the `cryptography.hazmat.primitives.ciphers.algorithms` module.

I have proposed a PR to the cryptography library to hopefully make the deprecation message clearer:

https://github.com/pyca/cryptography/pull/11407

**Comment by ErickAlejo** [comment]
> As a note. The TripleDES cipher is not being removed from the cryptography library. It is just being removed from the `cryptography.hazmat.primitives.ciphers.algorithms` module.
> 
> I have proposed a PR to the cryptography library to hopefully make the deprecation message clearer:
> 
> [pyca/cryptography#11407](https://github.com/pyca/cryptography/pull/11407)

Oooh that's good, all that's left to do is wait for the merge...

**Comment by toralf** [comment]
My Ansible playbook spews it 2x in a row at each start :-/

**Comment by igoruhhh** [comment]
to try changed: 
for paramiko/pkey.py
use:
 import from cryptography.hazmat.decrepit.ciphers.algorithms import TripleDES and changed 
# known encryption types for private key files:
CIPHER_TABLE = {
    "DES-EDE3-CBC": {
        "cipher": TripleDES,  <<<---------
        "keysize": 24,
        "blocksize": 8,
        "mode": modes.CBC,
    },
}
    --------------------------------
    for /transport.py 
    use:
    from cryptography.hazmat.decrepit.ciphers.algorithms import TripleDES

_cipher_info = {
    ...
    "3des-cbc": {
        "class": TripleDES,  <<< -------
        "mode": modes.CBC,
        "block-size": 8,
        "key-size": 24,
    },
}

**Comment by bitprophet** [comment]
Thanks all! The linked fix has been merged and new Paramiko 3.3/3.4 line releases will be on PyPI shortly.

**Comment by Wedge009** [comment]
I encourage people to read the entire issue, and particularly take note that a proper resolution has been merged, instead of using work-arounds.

**Comment by ohrl** [comment]
just comments these lines in ~/cryptography/hazmat/primitives/ciphers/algorithms.py:91
```
# utils.deprecated(
#     TripleDES,
#     __name__,
#     "TripleDES has been moved to "
#     "cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and "
#     "will be removed from this module in 48.0.0.",
#     utils.DeprecatedIn43,
#     name="TripleDES",
# )
```

**Comment by HydraDragonAntivirus** [comment]
Same issue here with different message: ```
 C:\Users\victim\AppData\Local\Programs\Python\Python312\Lib\site-packages\scapy\layers\ipsec.py:512: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from this module in 48.0.0.
  cipher=algorithms.TripleDES,
c:\Users\victim\AppData\Local\Programs\Python\Python312\Lib\site-packages\scapy\layers\ipsec.py:516: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from this module in 48.0.0.
  cipher=algorithms.TripleDES,
```

**Comment by schhabra-cinemo** [comment]
It's resolved in latest version paramiko==3.4.1

**Comment by bitprophet** [comment]
Yes, sorry for not being explicit, Paramiko 3.3.2 and 3.4.1 both include the fix for this warnings issue.
